goalie/interpolation.py

```diff
@@ -1,72 +1,64 @@
 """
 Driver functions for mesh-to-mesh data transfer.
 """
-from .utility import assemble_mass_matrix
+from .utility import assemble_mass_matrix, cofunction2function, function2cofunction
 import firedrake
+from firedrake.functionspaceimpl import WithGeometry
 from firedrake.petsc import PETSc
 from petsc4py import PETSc as petsc4py
-from typing import Union
-import ufl
 
 
 __all__ = ["project"]
 
 
-def project(
-    source: firedrake.Function,
-    target_space: Union[firedrake.FunctionSpace, firedrake.Function],
-    adjoint: bool = False,
-    **kwargs,
-) -> firedrake.Function:
-    """
+def project(source, target_space, **kwargs):
+    r"""
     Overload :func:`firedrake.projection.project` to account for the case of two mixed
-    function spaces defined on different meshes and for the adjoint projection operator.
+    function spaces defined on different meshes and for the adjoint projection operator when applied to :class:`firedrake.cofunction.Cofunction`\s.
 
     Extra keyword arguments are passed to :func:`firedrake.projection.project`.
 
-    :arg source: the :class:`firedrake.function.Function` to be projected
+    :arg source: the :class:`firedrake.function.Function` or
+        :class:`firedrake.cofunction.Cofunction` to be projected
     :arg target_space: the :class:`firedrake.functionspaceimpl.FunctionSpace` which we
-        seek to project into
-    :kwarg adjoint: apply the transposed projection operator?
+        seek to project into, or the :class:`firedrake.function.Function` or
+        :class:`firedrake.cofunction.Cofunction` to use as the target
     """
-    if not isinstance(source, firedrake.Function):
-        raise NotImplementedError("Can only currently project Functions.")  # TODO
-    Vs = source.function_space()
-    if isinstance(target_space, firedrake.Function):
+    if not isinstance(source, (firedrake.Function, firedrake.Cofunction)):
+        raise NotImplementedError(
+            "Can only currently project Functions and Cofunctions."
+        )
+
+    # If the input is a Cofunction then record this, map to a Function for the
+    # projection and then map back to a Cofunction afterwards
```

Comment:
I was worried when I read this comment - as it seems to suggest you convert to a function, then do a normal projection and then convert back - but you are actually doing an adjoint projection

Reply:
Reworded in 1b825f0.